### PR TITLE
Added regex check to prevent certain special characers from being placed in index properties fields

### PR DESCRIPTION
**Pull request by ryanbogan**
Signed-off-by: Ryan Bogan <rbogan@amazon.com>

### Description
Previously, the API would allow indices with special characters to be added.  Now, a MapperParsingException will be thrown when certain special characters are part of the properties field.

```
{
  "error" : {
    "root_cause" : [
      {
        "type" : "mapper_parsing_exception",
        "reason" : "Properties cannot contain special characters: Test<Field>Name?*#\\"
      }
    ],
    "type" : "mapper_parsing_exception",
    "reason" : "Properties cannot contain special characters: Test<Field>Name?*#\\"
  },
  "status" : 400
}
```
 
### Check List
- [X ] New functionality includes testing.
  - [X ] All tests pass
- [X ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java [hunk 1]
```diff
@@ -349,6 +351,11 @@ protected static void parseProperties(ObjectMapper.Builder objBuilder, Map<Strin
             while (iterator.hasNext()) {
                 Map.Entry<String, Object> entry = iterator.next();
                 String fieldName = entry.getKey();
+                Pattern p = Pattern.compile("[^a-zA-Z0-9_.@ $-]");
```
Comment (dblock) [inline server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:354]: Compile the pattern outside of the iteration?

**Review by dblock — CHANGES_REQUESTED** [review]

**Comment by dblock** [comment]
I think this is 3.0 (main) only, otherwise it's an API change. Any objections?

File: server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java (hunk 1 above)
Comment (ryanbogan) [reply server/src/main/java/org/opensearch/index/mapper/ObjectMapper.java:354]: I'll add that now

**Comment by dblock** [comment]
I googled https://opensearch.org/docs/latest/opensearch/rest-api/index-apis/create-index/#index-naming-restrictions - where do we enforce this behavior? Any changes should go there probably.

**Comment by CEHENKLE** [comment]
Hey @dblock I think what @ryanbogan is trying to sanitize inputs to reduce the risk of Cross-Site Scripting attacks (which is generally a Good Thing To Do(tm).  We might want to narrow the blast radius with something aimed specifically at tags like https://gist.github.com/pseudosavant/0187ea97d5fd87ce78bb?permalink_comment_id=3896888#gistcomment-3896888.  

If you've got an even better way to achieve that objective, definitely all for it!

I agree that this is a breaking change, and we'll need to push it to 3.0.  Tagging @elfisher for assistance on rollout plan.

**Comment by dblock** [comment]
Thanks @CEHENKLE, makes sense. See my questions above.

**Comment by macohen** [comment]
Would it be possible to make this optional and off by default in 2.x with documentation as to why it is recommended to turn on the sanitization and then turn it on by default in 3.x? Agreed that other languages need to be considered here, too.

**Comment by CEHENKLE** [comment]
@dblock I'm in favor of scoping the regex down just to catch javascript tags.  WDYT?

**Comment by dblock** [comment]
I am against this change. It's both breaking, and gives an impression of safety without actually making anything safe. Clients must sanitize user data being display anyway, so why is this change needed?

**Comment by CEHENKLE** [comment]
Okay, a couple more thoughts:
1) @dblock is right (Man, I hate tying that ;) ;) ;) ) -- this is definitely a breaking change, so there's no way we'd put it in before 3.0 .  This gives us some time to think it through.
2) @dblock is also right (auuugh, such torture ;) ;) ;) ) that what the "right thing to do" is something we'd have to put a fair amount of thought into.  What and why we sanitize deserves its own proposal and discussion, because there are tons of ramifications, particularly around migration.

I'd be interested to hear other voices in the conversation, so I'd like to leave this open for a bit.

-- Do you think it's the responsibility of the database to sanitize inputs?  
-- OpenSearch already puts restrictions on index names.  Should we also put restrictions on property names to be more cohesive?
-- If we put new restrictions in, what should be the migration path for existing property values that now fail checks?

Thanks,
/C

**Comment by dblock** [comment]
Note that we could add a feature in which users can optionally specify restrictions on the properties field character set, and ship that in an x.y release. It could become a default in a major version if it gets adopted widely.

**Pull request closed without merging**